src/plugins/intel_gpu/src/kernel_selector/cl_kernels/lora_ref.cl

```diff
@@ -0,0 +1,182 @@
+                 const __global STATE_TYPE* state_alpha,
+                 const __global STATE_TYPE* state_b,
+                       __global OUTPUT_TYPE* output)
+{
+    const uint K = INPUT1_SIZE_Y;
+
+    uint bf = get_group_id(0);
+    uint yx = get_group_id(1);
+
+    uint local_id = get_local_id(1);
+
+    ACCUMULATOR_TYPE acc = ACCUMULATOR_VAL_ZERO;
+
+    for (uint ki = 0; ki < K; ++ki) {
+        uint lora_input_idx = bf * K + ki;
+        uint state_a_idx = ki * LORA_RANK + local_id;
+
+        acc = mad(TO_ACCUMULATOR_TYPE(lora_input[lora_input_idx]),
+                  TO_ACCUMULATOR_TYPE(state_a[state_a_idx]),
+                  acc);
+    }
+
+    acc *= TO_ACCUMULATOR_TYPE(state_alpha[local_id]);
+
+    __local ACCUMULATOR_TYPE tmp_buf[MAX_LORA_RANK];
+    tmp_buf[local_id] = acc;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    ACCUMULATOR_TYPE final_acc = ACCUMULATOR_VAL_ZERO;
+    uint new_yx = yx * LORA_RANK + local_id;
+
+    if (new_yx >= INPUT0_SIZE_Y * INPUT0_SIZE_X) {
+        return;
+    }
+
+    for (uint ki = 0; ki < LORA_RANK; ++ki) {
+        uint state_b_idx = ki * INPUT0_SIZE_Y * INPUT0_SIZE_X + new_yx;
+
+        final_acc = mad(tmp_buf[ki], TO_ACCUMULATOR_TYPE(state_b[state_b_idx]), final_acc);
+    }
+    uint output_idx = bf * INPUT0_SIZE_Y * INPUT0_SIZE_X + new_yx;
+    output[output_idx] = TO_OUTPUT_TYPE(final_acc) + main_input[output_idx];
+}
+
+#else
+
+KERNEL(lora_ref)(OPTIONAL_SHAPE_INFO_ARG
+                 const __global OUTPUT_TYPE* main_input,
+                 const __global INPUT1_TYPE* lora_input,
+                 const __global STATE_TYPE* state_a_0,
+                 const __global STATE_TYPE* state_alpha_0,
+                 const __global STATE_TYPE* state_b_0,
+                 const __global STATE_TYPE* state_a_1,
+                 const __global STATE_TYPE* state_alpha_1,
+                 const __global STATE_TYPE* state_b_1,
+#if LORA_COUNT == 3
+                 const __global STATE_TYPE* state_a_2,
+                 const __global STATE_TYPE* state_alpha_2,
+                 const __global STATE_TYPE* state_b_2,
+#endif
+                       __global OUTPUT_TYPE* output)
+{
+    const uint K = INPUT1_SIZE_Y;
+
+    const uint bf = get_group_id(0);
+    const uint yx = get_group_id(1);
+    const uint local_id = get_local_id(1);
+
+    ACCUMULATOR_TYPE acc0 = ACCUMULATOR_VAL_ZERO;
+    ACCUMULATOR_TYPE acc1 = ACCUMULATOR_VAL_ZERO;
+#if LORA_COUNT == 3
+    ACCUMULATOR_TYPE acc2 = ACCUMULATOR_VAL_ZERO;
+#endif
+
+    for (uint ki = 0; ki < K; ++ki) {
+        uint lora_idx = bf * K + ki;
+        uint state_idx = local_id * K + ki;
+
+        ACCUMULATOR_TYPE lora_val = TO_ACCUMULATOR_TYPE(lora_input[lora_idx]);
+
+        acc0 = mad(lora_val, TO_ACCUMULATOR_TYPE(state_a_0[state_idx]), acc0);
+        acc1 = mad(lora_val, TO_ACCUMULATOR_TYPE(state_a_1[state_idx]), acc1);
+#if LORA_COUNT == 3
+        acc2 = mad(lora_val, TO_ACCUMULATOR_TYPE(state_a_2[state_idx]), acc2);
+#endif
```

Comment:
lora_input is the output from FCHorizontalFuse. It looks A gemm of each lora should only operate a part of it by splitting it horizontally. Any misunderstanding here?

Reply:
`main_input` is the output from FCHorizontalFuse, `lora_input` is the output from FCHorizontalFuse input

Reply:
oh, yeh, I see.